src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -786,10 +786,18 @@ private TypeSymbol BindVariableType(CSharpSyntaxNode declarationNode, Diagnostic
             // might own nested scope.
             bool hasErrors = localSymbol.ScopeBinder.ValidateDeclarationNameConflictsInScope(localSymbol, diagnostics);
 
-            var containingMethod = this.ContainingMemberOrLambda as MethodSymbol;
-            if (containingMethod != null && containingMethod.IsAsync && localSymbol.RefKind != RefKind.None)
+            if (localSymbol.RefKind == RefKind.In)
             {
-                Error(diagnostics, ErrorCode.ERR_BadAsyncLocalType, declarator);
+                var refKeyword = typeSyntax.GetFirstToken();
+                diagnostics.Add(ErrorCode.ERR_UnexpectedToken, refKeyword.GetLocation(), refKeyword.ToString());
```

Comment:
ReadonlyRefReturn_UnexpectedBindTime

We allow `readonly ref` locals at parse time for error recovery purposes, but will fail at bind time. 

This is the same strategy as used for regular `ref`. We parse it as RefType in more places than where it is semantically allowed, but fail to bind with this error.

Also there is a possibility we would allow `readonly ref` locals. That is another reason for having the check at bind time.

---
In reply to: [102299405](https://github.com/dotnet/roslyn/pull/17223#discussion_r102299405) [](ancestors = 102299405)